modules/generative-openai/clients/openai.go

```diff
@@ -71,22 +73,48 @@ func (v *openai) GenerateAllResults(ctx context.Context, textProperties []map[st
 func (v *openai) Generate(ctx context.Context, cfg moduletools.ClassConfig, prompt string) (*ent.GenerateResult, error) {
 	settings := config.NewClassSettings(cfg)
 
-	body, err := json.Marshal(generateInput{
-		Prompt:           prompt,
-		Model:            settings.Model(),
-		MaxTokens:        settings.MaxTokens(),
-		Temperature:      settings.Temperature(),
-		FrequencyPenalty: settings.FrequencyPenalty(),
-		PresencePenalty:  settings.PresencePenalty(),
-		TopP:             settings.TopP(),
-	})
-	if err != nil {
-		return nil, errors.Wrapf(err, "marshal body")
+	var body []byte
+	var err error
+	var oaiUrl string
+	var input generateInput
+
+	if settings.IsLegacy() {
+		oaiUrl = path.Join(v.host, "/v1/completions")
+		if err != nil {
+			return nil, errors.Wrap(err, "join OpenAI API host and path")
+		}
+		input = generateInput{
+			Prompt:           prompt,
+			Model:            settings.Model(),
+			MaxTokens:        settings.MaxTokens(),
+			Temperature:      settings.Temperature(),
+			FrequencyPenalty: settings.FrequencyPenalty(),
+			PresencePenalty:  settings.PresencePenalty(),
+			TopP:             settings.TopP(),
+		}
+	} else {
+		oaiUrl, err = url.JoinPath(v.host, v.path)
+		if err != nil {
+			return nil, errors.Wrap(err, "join OpenAI API host and path")
+		}
+		tokens := determineTokens(settings.GetMaxTokensForModel(settings.Model()), settings.MaxTokens(), prompt)
+		input = generateInput{
+			Messages: []message{{
+				Role:    "user",
+				Content: prompt,
+			}},
+			Model:            settings.Model(),
+			MaxTokens:        float64(tokens),
+			Temperature:      settings.Temperature(),
+			FrequencyPenalty: settings.FrequencyPenalty(),
+			PresencePenalty:  settings.PresencePenalty(),
+			TopP:             settings.TopP(),
+		}
 	}
 
-	oaiUrl, err := url.JoinPath(v.host, v.path)
+	body, err = json.Marshal(input)
```

Comment:
after changes to what is being done inside `if statements` you can do here:  `body, err := json.Marshal(input)` and remove `var body []byte` and `var error`